apps/src/code-studio/components/progress/Announcements.jsx

```diff
@@ -0,0 +1,64 @@
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import {
+  announcementShape,
+  VisibilityType
+} from '@cdo/apps/code-studio/announcementsRedux';
+import Notification from '@cdo/apps/templates/Notification';
+import i18n from '@cdo/locale';
+import {ViewType} from '@cdo/apps/code-studio/viewAsRedux';
+
+export default class Announcements extends Component {
+  static propTypes = {
+    announcements: PropTypes.arrayOf(announcementShape).isRequired,
+    width: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
+    viewAs: PropTypes.oneOf(Object.values(ViewType)).isRequired
+  };
+
+  /*
+  Processes all of the announcements and determines if they should be shown based
+  on the their visibility setting and the current view. For example a teacher should only see
+  announcements for Teacher-only or Teacher and Student.
+  Also defaults to old announcements without a visibility to be Teacher-only.
+ */
+  filteredAnnouncements = () => {
+    const currentAnnouncements = [];
+    const currentView = this.props.viewAs;
+    this.props.announcements.forEach(element => {
+      if (element.visibility === VisibilityType.teacherAndStudent) {
+        currentAnnouncements.push(element);
+      } else if (
+        currentView === 'Teacher' &&
+        (element.visibility === VisibilityType.teacher ||
+          element.visibility === undefined)
+      ) {
+        currentAnnouncements.push(element);
+      } else if (
+        currentView === 'Student' &&
+        element.visibility === VisibilityType.student
+      ) {
+        currentAnnouncements.push(element);
+      }
+    });
+    return currentAnnouncements;
```

Comment:
I would suggest extracting a helper method here and then using `filter`. the benefit is that it separates the business logic ("is it visible") from what you are doing (constructing the array). there are many options but here is one way:
```
isVisible = (currentView, element) => (
  element.visibility === VisibilityType.teacherAndStudent ||
  currentView === 'Teacher' && (element.visibility === VisibilityType.teacher || element.visibility === undefined) ||
  currentView === 'Student' && element.visibility === VisibilityType.student
)
```
you could also use a switch statement if you prefer how that looks. either way, this method is then reduced to:
```suggestion
    const currentView = this.props.viewAs;
    return this.props.announcements.filter(element => this.isVisible(currentView, element));
```